### PR TITLE
Tick methods implementation in precompile

### PR DESCRIPTION
**Pull request by shubhamgoyal42**

### CODE REVIEW

File: precompile/contracts/juror/logic.go [hunk 1]
```diff
@@ -386,6 +387,114 @@ func validateExecuteIOCOrder(bibliophile b.BibliophileClient, order *orderbook.I
...
 
+func GetPrevTick(bibliophile b.BibliophileClient, input GetPrevTickInput) (*big.Int, error) {
+	if input.Tick.Sign() == 0 {
+		return nil, errors.New("tick price cannot be zero")
+	}
+	if input.IsBid {
+		bidsHead := bibliophile.GetBidsHead(input.Amm)
+		if input.Tick.Cmp(bidsHead) != -1 {
+			return nil, fmt.Errorf("tick %d is greater than or equal to bidsHead %d", input.Tick, bidsHead)
+		}
```
Comment (fdgod09) [inline precompile/contracts/juror/logic.go:398]: We dont need this check. It is already covered in contract and logic will return bidsHead if called anyway.
Same for asks

File: precompile/contracts/juror/logic.go [hunk 2]
```diff
@@ -386,6 +387,114 @@ func validateExecuteIOCOrder(bibliophile b.BibliophileClient, order *orderbook.I
 	}, nil
 }
 
+func GetPrevTick(bibliophile b.BibliophileClient, input GetPrevTickInput) (*big.Int, error) {
+	if input.Tick.Sign() == 0 {
```
Comment (fdgod09) [inline precompile/contracts/juror/logic.go:391]: Dont need to check if input == 0. Logic can handle it. We allow 0 price orders

File: precompile/contracts/juror/logic.go [hunk 3]
```diff
@@ -386,6 +388,114 @@ func validateExecuteIOCOrder(bibliophile b.BibliophileClient, order *orderbook.I
...
+		}
+		accNotional = accumulator
+		accBaseQ.Add(accBaseQ, amount)
+		nextTick := bibliophile.GetNextBidPrice(ammAddress, tick)
+		tick = nextTick
+	}
+	if tick.Sign() == 0 {
+		return big.NewInt(0).Div(big.NewInt(0).Mul(accNotional, big.NewInt(1e18)), accBaseQ)
+	}
+	baseQAtTick := new(big.Int).Div(new(big.Int).Mul(new(big.Int).Sub(impactMarginNotional, accNotional), big.NewInt(1e6)), tick)
```
Comment (atvanguard) [inline precompile/contracts/juror/logic.go:452]: should be multiplied by 1e18. Same for `SampleImpactAsk`

File: precompile/contracts/bibliophile/client.go [hunk 4]
```diff
@@ -8,35 +8,44 @@ import (
...
-	GetUpperAndLowerBoundForMarket(marketId int64) (*big.Int, *big.Int)
 	GetLongOpenOrdersAmount(trader common.Address, ammIndex *big.Int) *big.Int
 	GetShortOpenOrdersAmount(trader common.Address, ammIndex *big.Int) *big.Int
 	GetReduceOnlyAmount(trader common.Address, ammIndex *big.Int) *big.Int
-
-	// Misc
 	IsTradingAuthority(senderOrSigner, trader common.Address) bool
-
+	GetMinAllowableMarginOB() *big.Int
+	GetTakerFeeOB() *big.Int
```
Comment (atvanguard) [inline precompile/contracts/bibliophile/client.go:26]: Curious why create 2 separate instances to read from OB and CH.

File: precompile/contracts/bibliophile/margin_account.go [hunk 5]
```diff
@@ -21,6 +22,11 @@ func GetNormalizedMargin(stateDB contract.StateDB, trader common.Address) *big.I
 	return getMargin(stateDB, big.NewInt(0), trader)
 }
 
+// GetMinAllowableMarginMA returns the minimum allowable margin for a trader set in MarginAccount
+func getMinAllowableMarginMA(stateDB contract.StateDB) *big.Int {
```
Comment (atvanguard) [inline precompile/contracts/bibliophile/margin_account.go:26]: why do we need 3 flavours for getting the same thing? `OB, CH, MA`

**Review by atvanguard — CHANGES_REQUESTED** [review]

File: precompile/contracts/juror/logic.go [hunk 6]
```diff
@@ -449,6 +559,10 @@ func validatePlaceLimitOrderV2(bibliophile b.BibliophileClient, order ILimitOrde
 		errorString = err.Error()
 		return
 	}
+	if order.Trader != trader && !bibliophile.IsTradingAuthority(order.Trader, trader) {
```
[the remark this answers is not in the document]
Comment (fdgod09) [reply precompile/contracts/juror/logic.go]: yes. tested it with unit tests

File: precompile/contracts/juror/logic.go (hunk 2 above)
Comment (fdgod09) [reply precompile/contracts/juror/logic.go:392]: we dont allow 0 price orders

**Review by atvanguard — APPROVED** [review]

**Pull request merged**